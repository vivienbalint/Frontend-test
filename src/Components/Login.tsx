import React, { useState } from "react";
import { Form, Button, Alert } from "react-bootstrap";
import api from "../api/api.user";
import AuthContext from "../auth/auth.context";
import { useHistory } from "react-router";

const Login = () => {
  let history = useHistory();
  let regex = /\[|]|"/g;

  const [{ email, password }, setCredentials] = useState({
    email: "",
    password: "",
  });
  const [error, setError] = useState("");

  const { dispatch } = React.useContext(AuthContext);

  const handleLogin = async (e: React.FormEvent) => {
    e.preventDefault();
    const { user, error } = await api.loginUser({ email, password });
    if (error) {
      setError(error);
      dispatch({ type: "LOGOUT" });
    } else if (user) {
      const userDataResponse = {
        user,
        token: user.token,
      };
      history.push("/user-page");

      dispatch({ type: "LOGIN", payload: userDataResponse });
    }
  };

  return (
    <div className="box">
      {error.length > 0 && (
        <Alert variant={"danger"}>
          <Alert.Heading>Error:</Alert.Heading>
          <p>{error.replace(regex, "")}</p>
        </Alert>
      )}
      <Form onSubmit={handleLogin} className="mx-auto form-container">
        <h2 className="form-title">Login</h2>
        <Form.Group className="mb-3">
          <Form.Label column="lg">Email address</Form.Label>
          <Form.Control
            size="lg"
            type="email"
            placeholder="Enter email address"
            value={email}
            onChange={(e) =>
              setCredentials({
                email: e.target.value,
                password,
              })
            }
          />
        </Form.Group>
        <Form.Group className="mb-3">
          <Form.Label column="lg">Password</Form.Label>
          <Form.Control
            size="lg"
            type="password"
            autoComplete="true"
            placeholder="Password"
            value={password}
            onChange={(e) =>
              setCredentials({
                email,
                password: e.target.value,
              })
            }
          />
        </Form.Group>
        <Button variant="primary" type="submit">
          Login
        </Button>
      </Form>
    </div>
  );
};

export default Login;
