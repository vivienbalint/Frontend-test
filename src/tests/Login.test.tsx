import React from "react";
import { mount, shallow } from "enzyme";
import Login from "../Components/Login";
import "../setupTests";

const dispatch = jest.fn();
React.useContext = (() => dispatch) as <T>(context: React.Context<T>) => T;

describe("Login", () => {
  const wrapper = mount(<Login />);

  it("should run without crashing", () => {
    shallow(<Login />);
  });

  it("should match the snapshot", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("should have an email field", () => {
    expect(wrapper.find('input[type="email"]').length).toEqual(1);
  });

  it("should have proper props for email field", () => {
    expect(wrapper.find('input[type="email"]').props()).toEqual({
      className: "form-control form-control-lg",
      id: undefined,
      readOnly: undefined,
      size: undefined,
      type: "email",
      placeholder: "Enter email address",
      value: expect.any(String),
      onChange: expect.any(Function),
    });
  });

  it("should have a password field", () => {
    expect(wrapper.find('input[type="password"]').length).toEqual(1);
  });

  it("should have proper props for password field", () => {
    expect(wrapper.find('input[type="password"]').props()).toEqual({
      className: "form-control form-control-lg",
      id: undefined,
      readOnly: undefined,
      size: undefined,
      type: "password",
      placeholder: "Password",
      value: expect.any(String),
      onChange: expect.any(Function),
      autoComplete: "true",
    });
  });

  it("should have a submit button", () => {
    expect(wrapper.find('button[type="submit"]').length).toEqual(1);
  });

  it("should have proper props for submit button", () => {
    expect(wrapper.find('button[type="submit"]').props()).toEqual({
      className: "btn btn-primary",
      disabled: false,
      type: "submit",
      children: "Login",
    });
  });
});
