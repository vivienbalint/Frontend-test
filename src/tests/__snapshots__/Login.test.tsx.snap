// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login should match the snapshot 1`] = `"<div class=\\"box\\"><form class=\\"mx-auto form-container\\"><h2 class=\\"form-title\\">Login</h2><div class=\\"mb-3\\"><label class=\\"form-label col-form-label col-form-label-lg col\\">Email address</label><input placeholder=\\"Enter email address\\" type=\\"email\\" class=\\"form-control form-control-lg\\" value=\\"\\"></div><div class=\\"mb-3\\"><label class=\\"form-label col-form-label col-form-label-lg col\\">Password</label><input autocomplete=\\"true\\" placeholder=\\"Password\\" type=\\"password\\" class=\\"form-control form-control-lg\\" value=\\"\\"></div><button type=\\"submit\\" class=\\"btn btn-primary\\">Login</button></form></div>"`;
